Reformatting and migrate raster to terra (#15)
GeoScripting-WUR/IntroToRaster: index.Rmd

# Overview of the *terra* package
The *raster* package used to be the reference R package for raster processing, with Robert J. Hijmans as its the original developer. The introduction of the *raster* package to R was a revolution for geo-processing and analysis using R. The *raster* package is now deprecated, as Robert Hijmans has developed a successor to it called *[terra](https://cran.r-project.org/web/packages/terra/)* which is both simpler and much faster, as it's rewritten in C++.

Among other things the *terra* package allows to:

* Read and write raster data of most commonly used formats.
* Perform most raster operations, such as creation of raster objects, performing spatial/geometric operations (re-projections, resampling, etc), filtering and raster calculations.
* Work on large raster datasets thanks to its built-in block processing functionalities.
* Perform fast operations thanks to optimized back-end C++ code.
* Visualize and interact with the data.
* etc...

```{block, type="alert alert-info"}

> **Tip**: Check the [home page](http://cran.r-project.org/web/packages/terra/) of the *terra* package. The package is extremely well documented, including vignettes and demos. See also the [reference manual](https://cran.r-project.org/web/packages/terra/terra.pdf) there.

Reviewer: It's useful to mention this terra reference: https://rspatial.org/spatial/index.html

```

## Explore the terra objects

> The terra package produces and uses R objects of two main classes. **SpatRaster** and **SpatVector**. A SpatRaster represents a spatially referenced surface divided into three dimensional cells (rows, columns, and layers). A Spatvector represents geometries as well as attributes (variables) describing the geometries.

Reviewer: ```suggestion
The terra package produces and uses R objects of two main classes: **SpatRaster** and **SpatVector**. A SpatRaster represents a spatially referenced surface divided into three dimensional cells (rows, columns, and layers). A SpatVector represents geometries as well as attributes (variables) describing the geometries. Note: we will be using *terra* only for raster processing, as you will see in subsequent tutorials.
```

Let's take a look into the structure of a SpatRaster.
```{r}
# Generate a SpatRaster
r <- rast(ncol = 40, nrow = 20)
class(r)

# Simply typing the object name displays its general properties / metadata
r
```

> From the metadata displayed above, we can see that the SpatRaster contains all the properties that geo-data should have; that is to say a projection, an extent and a pixel resolution.

Reviewer: Technically:
```suggestion
From the metadata displayed above, we can see that the SpatRaster contains all the properties that geo-data should have; that is to say a coordinate reference system, an extent and a pixel size.
```

Writing a SpatRaster to file is achieved using the `writeRaster()` function.

To illustrate the reading and writing of raster files, we will use data subsets that we have prepared for the course and need to be downloaded from the repository. For that, first make sure your working directory is set properly. Then run the following line; it will handle the download:

```{r, eval=TRUE}
# Start by making sure that your working directory is properly set
# If not you can set it using setwd()
getwd()

# Create data directory if it does not yet exist
if (!dir.exists("data")) {
  dir.create("data")
}

# Download the data
# In case the download code doesn't work, use method = 'wget'
download.file(url = 'https://github.com/GeoScripting-WUR/IntroToRaster/releases/download/tahiti/gewata.zip', destfile = 'data/gewata.zip')

# Unpack the archive
unzip('data/gewata.zip', exdir = "data")
```

> *Gewata* is the name of the data set added. It is a multi-layer GeoTIFF object, and its file name is *LE71700552001036SGS00\_SR\_Gewata\_INT1U.tif*, informing us that this is a subset from a scene acquired by the Landsat 7 sensor. Let's not worry about the region that the data covers for now, we will find a nice way to discover that later on in the tutorial.

Reviewer: ```suggestion
*Gewata* is the name of the dataset that we just downloaded. It is a multi-layer GeoTIFF object, and its file name is *LE71700552001036SGS00\_SR\_Gewata\_INT1U.tif*, informing us that this is a subset from a scene acquired by the Landsat 7 ETM+ sensor. Let's not worry about the region that the data covers for now, we will find a nice way to discover that later on in the tutorial.
```

```{r}
gewata
```
The metadata above informs us that the gewata object is a relatively small (593x653 pixels) SpatRaster with 6 layers.

## Data type is (still) important
When writing files to disk using `writeRaster()` or the `filename =` argument in most raster processing functions, you are able set an appropriate data type. Using the `datatype =` argument, you can save some precious disk space compared to the default datatype, and thus increase read and write speed.

## Geoprocessing: in memory vs. on disk
When looking at the documentation of most functions of the *terra* package, you will notice that the list of arguments is almost always ended by `...`. These 'three dots' are called an ellipsis; it means that extra arguments can be passed to the function. Often these arguments are those that can be passed to the `writeRaster()` function; meaning that most geoprocessing functions are able to write their output directly to file, on disk. This reduces the number of steps and is always a good consideration when working with big raster objects that tend to overload the memory if not written directly to file.

## Cropping a SpatRaster

> `crop()` is the terra package function that allows you to crop data to smaller spatial extents. A great advantage of the crop function is that it accepts almost all spatial object classes in R as its `extent` input argument. But the `extent` argument also simply accepts objects of class `extent`. One way of obtaining such an extent object interactively is by using the `draw()` function. In the example below, we will manually draw a regular extent that we will use later to crop the *gewata* SpatRaster.

Reviewer: The function doesn't have an `extent` input argument any more. The `y` argument accepts `SpatExtent`, not `extent`.

```{r, eval=FALSE}
# Plot the first layer of the SpatRaster
plot(gewata, 1)
e <- draw()
```

> Now you have to define a rectangular bounding box that will define the spatial extent of the extent object. Click twice, for the two opposite corners of the rectangle. Now we can crop the data following the boundaries of this extent.

Reviewer: It may be good to add a note here that R will wait until you finish drawing before it will let you run any more code. A lot of students think that their R crashes or is very slow when they do draw(), but it's just waiting for them to finish drawing. (I remember a case where a student said "I'm waiting for R to run my code!", to which I replied "and R is waiting for you to draw an extent!" :) )

```{r, eval=FALSE}
# Crop gewata using e
gewataSub <- crop(gewata, e)

# Visualize the new cropped object
plot(gewataSub, 1)
```

You should see on the resulting plot that the original image has been cropped.

## Creating layer stacks

> To end this section on general files and raster object manipulations, we will see more in-depth how multi-layer objects can be created from single-layer objects. The object created as part of the example below is the same that we will use later in the course to perform time series analysis on raster objects. It is composed of NDVI layers derived from Landsat acquisitions at different dates. The objective is therefore to create a multi-layer NDVI object, for which each layer corresponds to a different date. But first we need to fetch the data, similarly to how we did it for the *gewata* dataset.

Reviewer: Hmm, do we actually use the tura dataset anywhere else in the course, as this paragraph states? I don't remember that...

Author: No, it used to be in the course, it's now in Demo 2: https://geoscripting-wur.github.io/Scripting4Geo/#Demo_2__Visualize_the_temporal_profile_of_a_pixel. I'll remove the reference from this text.

getwd()

# Download and unzip the data
download.file(url = 'https://github.com/GeoScripting-WUR/IntroToRaster/releases/download/tahiti/tura.zip', destfile = 'data/tura.zip')
unzip(zipfile = 'data/tura.zip', exdir = 'data')

# Retrieve the content of the tura sub-directory
list <- list.files(path = 'data/tura/', full.names = TRUE)
```

The object `list` contains the file names of all the single layers we have to stack. Let's open the first one to visualize it.

```{r, eval=FALSE}
plot(rast(list[1]))
```

We see an NDVI layer, with the clouds masked out. Now let's create a multi-layer SpatRaster.

```{r, eval=FALSE}
turaStack <- rast(list)
turaStack
```

Now that we have our SpatRaster with 166 layers in memory, let's write it to disk using the `writeRaster()` function. Note that we adjust the layer names to the original file names (in which information on dates is written). Also note that the data range is comprised between -10000 and +10000, therefore such a file can be stored as signed 2 byte integer (INT2S).

```{r, eval=FALSE}
# Create output directory if it does not yet exist
if (!dir.exists("output")) {
  dir.create("output")
}

# Write this file to the 

> writeRaster(x = turaStack, filename = 'output/turaStack.tif', names = list.files(path = 'data/tura/'), datatype = "INT2S")

Reviewer: Alternatively `names = basename(list)`

Author: We use list.files() in 205 as well, which also seems more intuïtive tbh

Reviewer: Yes, it's just a bit odd that the `list` object (which is not a `list` ;) ) is then not so useful :) But it's not a big deal.

```

Now this object is stored on your computer, ready to be archived for later use.

# Simple raster arithmetic
## Adding, subtracting, multiplying and dividing SpatRasters

> Performing simple raster operations with a SpatRaster is fairly easy. For instance, if you want to subtract two SpatRasters of same extent, `r1` and `r2`; simply doing `r1 - r2` will give the expected output, which is, every pixel value of `r2` will be subtracted from the matching pixel value of `r1`. These types of pixel-based operations almost always require a set of conditions to be met in order to be executed; the two SpatRasters need to be identical in term of extent, resolution, projection, etc.

Reviewer: ```suggestion
Performing simple raster operations with a SpatRaster is fairly easy. For instance, if you want to subtract two SpatRasters of same extent, `r1` and `r2`; simply doing `r1 - r2` will give the expected output, which is, every pixel value of `r2` will be subtracted from the matching pixel value of `r1`. These types of pixel-based operations almost always require a set of conditions to be met in order to be executed; the two SpatRasters need to be identical in term of extent, pixel size, coordinate reference system, etc.
```

## Subsetting layers from SpatRaster

> Different spectral bands of a same satellite scene are often stored in multi-layer objects. This means that you will very likely import them in your R working environment as one multi-layer SpatRaster. As a consequence, to perform calculations between these bands, you will have to write an expression referring to individual layers of the object. Referring to individual layers in a SpatRaster can be done by using double square brackets `[[]]`.

Reviewer: ```suggestion
Different spectral bands of the same satellite scene are often stored in multi-layer objects. This means that you will very likely import them in your R working environment as one multi-layer SpatRaster. As a consequence, to perform calculations between these bands, you will have to write an expression referring to individual layers of the object. Referring to individual layers in a SpatRaster can be done by using double square brackets `[[]]`.
```

```{r}
all.equal(ndvi, ndvi2)
```

Here we see that only the name of `ndvi2` is different. This is because for the first method, which resulted in `ndvi`, the layer name was copied from `gewata[[4]]`, whereas this was not the case for the second approach.

## Simple raster statistics
The `global()` function allows the extraction of basic global statistics from an entire SpatRaster, such as the mean value, maximum value, the range of values, or the number of NA cells. `global` will return a dataframe, from which we can consequently select the value we need. See also `?global` for more information and documentation. Let's try some of these options for our `ndvi` SpatRaster. 
```{r}
# Calculate the mean (this produces a dataframe)
mean <- global(ndvi, fun = 'mean')
mean

# Get only the value from the dataframe
mean$mean

# Calculate the amount of NA values (which should be zero in this case)
global(ndvi, fun = 'isNA')

# Calculate the amount of non-NA values
global(ndvi, fun = 'notNA')

# Which, as we have no NA values, should be equal to
ncell(ndvi)
```

## Reprojections

> The `project()` function allows re-projection of raster objects to any projection one can think of. It accepts the following formats to define coordinate reference systems: *WKT*, *PROJ.4* (e.g., +proj=longlat +datum=WGS84), or an *EPSG* code (e.g., "epsg:4326"). Note that the PROJ.4 notation has been deprecated, and you can only use it with the WGS84/NAD83 and NAD27 datums. Other datums are silently ignored.

Reviewer: ```suggestion
The `project()` function allows re-projection of raster objects to any projection one can think of. It accepts the following formats to define coordinate reference systems: *WKT*, *PROJ.4* (e.g., `+proj=longlat +datum=WGS84`), or an *EPSG* code (e.g., `epsg:4326`). Note that the PROJ.4 notation has been deprecated, and you can only use it with the WGS84/NAD83 and NAD27 datums. Other datums are silently ignored.
```

> A central place to search for projections is the spatial reference website ([http://spatialreference.org/](http://spatialreference.org/)), from this database you will be able to query almost any reference and retrieve it in any format, including its *EPSG* code. Well-Known Text (*WKT*) expressions are preferred for scientific correctness and lack of ambiguity.

Reviewer: ```suggestion
A central place to search for coordinate reference systems is the spatial reference website ([http://spatialreference.org/](http://spatialreference.org/)), from this database you will be able to query almost any reference and retrieve it in any format, including its *EPSG* code. Well-Known Text (*WKT*) expressions are preferred for scientific correctness and lack of ambiguity.
```

> Instead of specifying a projection to convert to, you can also provide `project()` with another SpatRaster, and it will convert your input to the projection of that SpatRaster.

Reviewer: ```suggestion
Instead of specifying a coordinate reference system to convert to, you can also provide `project()` with another SpatRaster, and it will convert your input to the CRS of that SpatRaster.
```

```{r}

> # One single line is sufficient to project any raster to any projection

Reviewer: ```suggestion
# One single line is sufficient to project any raster to any CRS
```

> ndviLL <- project(ndvi, '+proj=longlat')

Reviewer: We should probably encourage using EPSG codes instead of the deprecated proj.4.

```

Note that if re-projecting and mosaicking is really a large part of your project, you may want to consider using the `gdalwarp` command line utility ([gdalwarp](http://www.gdal.org/gdalwarp.html)) directly. The `gdalUtils` R package provides utilities to run GDAL commands from R, including `gdalwarp`, for reprojection, resampling and mosaicking.

## Exporting and inspecting

> By the way, we still don't know where this area is. In order to investigate that, we are going to try projecting it in QGIS. Let's write our NDVI layer in Lat/Long to a .tif file first.

Reviewer: ```suggestion
By the way, we still don't know where this area is. In order to investigate that, we are going to try displaying it in QGIS. Let's write our NDVI layer in Lat/Long to a .tif file first.
```

```{r, eval = FALSE}
# Since this function will write a file to your working directory
# you want to make sure that it is set where you want the file to be written
# It can be changed using setwd()
getwd()

# Note that we are using the filename argument, contained in the ellipsis (...) of
# the function, since we want to write the output directly to file.
writeRaster(x = ndviLL, filename = 'output/gewataNDVI.tif')
```

> Now open this file in QGIS (or ArcGIS), and add a base layer (in QGIS for instance OpenStreetMap, using XYZ tiles with the url `https://tile.openstreetmap.org/{z}/{x}/{y}.png`). Once we zoom out a bit, we see we are all the way in ... Ethiopia. More information will come later in the course about that specific area.

Reviewer: Not in ArcGIS, as it does not run in Linux, and that's what we are using in the course.

## Reprojections
The `project()` function allows re-projection of raster objects to any projection one can think of. It accepts the following formats to define coordinate reference systems: *WKT*, *PROJ.4* (e.g., `+proj=longlat +datum=WGS84`), or an *EPSG* code (e.g., `epsg:4326`). Note that the PROJ.4 notation has been deprecated, and you can only use it with the WGS84/NAD83 and NAD27 datums. Other datums are silently ignored.

A central place to search for coordinate reference systems is the spatial reference website ([http://spatialreference.org/](http://spatialreference.org/)), from this database you will be able to query almost any reference and retrieve it in any format, including its *EPSG* code. Well-Known Text (*WKT*) expressions are preferred for scientific correctness and lack of ambiguity.

Instead of specifying a coordinate reference system to convert to, you can also provide `project()` with another SpatRaster, and it will convert your input to the CRS of that SpatRaster.

```{r}
# One single line is sufficient to project any raster to any CRS

> ndviLL <- project(ndvi, 'epsg:28992')

Reviewer: The "LL" I think stands for Long Lat, but 28992 is RD New. I think we are supposed to use WGS84 here?

```{r}
# Perform value replacement
tahiti6[fmask != 0] <- NA
```

However, this is possible here because both objects are relatively small and the values can all be loaded in the computer memory without any risk of overloading it. When working with very large raster objects, you will very likely run into problems if you do that. It is then preferable, as presented earlier in this tutorial to use `app()`.

```{r}
# First define a value replacement function
cloud2NA <- function(x) {
    x[1:6][x[7] != 0] <- NA
    return(x)
}
```

The value replacement function takes one argument, `x`. `x` corresponds to a SpatRaster, where the 7th layer is our cloud mask.

```{r, fig.align='center'}
# Let's create a new 6 layers object since tahiti6 has now been masked already
tahiti6_2 <- subset(tahiti, 1:6)

# Apply the function on the two SpatRasters

> tahitiCloudFree <- app(x = c(tahiti6_2, fmask), fun = cloud2NA)

Reviewer: With `app()` this is actually much simpler: `c(tahiti6_2, fmask)` is simply `tahiti`. So we can skip line 417 altogether.

# Visualize the output
plotRGB(tahitiCloudFree, 3, 4, 5, stretch = "lin")
```

There are holes in the image, but at least the clouds are gone. We could use another image from another date, to create a composite image, but that is a little bit too much for today.

# Summary
Today you got a general introduction to the *terra* package, its basic functions, its object classes and methods. They can be categorized as follows:

## Terra classes

* `SpatRaster`: Single- or multi-layer object (raster).

> * `SpatVector`: Geometries object (vector).

Reviewer: ```suggestion
* `SpatRaster`: Single- or multi-layer raster object.
* `SpatVector`: Vector object.
```

## Terra classes

* `RasterLayer`: Single-layer object.
* `RasterStack` and `RasterBrick`: Multi-layer raster objects.
* `SpatRaster`: Single- or multi-layer object (raster).
* `SpatVector`: Geometries object (vector).

## Functions

### Read data

* `raster()`: Read a single-layer raster object written on disk, or read the first layer of a multi-layer object.
* `brick()`: Read a multi-layer raster object written on disk.

### Write data
### Reading and writing

* `writeRaster()`: Write a `RasterLayer`, `RasterBrick` or `RasterStack` to disk.
* `filename =` argument: Available for most functions of the *raster* package that produce raster objects, write directly the output of the function to disk.
* `rast()`: Read a raster object from disk.
* `writeRaster()`: Write a `SpatRaster` to disk.
* `filename =` argument: Available for most functions of the *terra* package that produce SpatRaster objects, write directly the output of the function to disk.

### Reformat data

* `crop()`: modify the extent of a `Raster*` object based on another spatial object or an extent object.
* `projectRaster()`: Reproject (and resample) a raster object to a desired coordinate reference system.

> * `stack()`: Assemble RasterLayers in a multi-layer object.

Reviewer: You should mention the use of `c()` in that case.

> * `crop()`: modify the extent of a `SpatRaster` based on another spatial object or an extent object.

Reviewer: ```suggestion
* `crop()`: modify the extent of a `SpatRaster` based on another spatial object or a `SpatExtent` object.
```

### Reformat data

* `crop()`: modify the extent of a SpatRaster based on another spatial object or a SpatExtent object.
* `project()`: Reproject (and resample) a SpatRaster to a desired coordinate reference system or reference SpatRaster.
* `subset()`: Create a subset of a SpatRaster.

> * `c()`: Combina multiple SpatRasters into a single SpatRaster.

Reviewer: ```suggestion
* `c()`: Combine multiple SpatRasters into a single SpatRaster.
```
